### PR TITLE
Controller to stream pool demand from a script

### PR DESCRIPTION
**Pull request by lsowa**
Adds a controller that sets a pool's demand from a live stream of values produced by an external script.

`StreamingController` spawns `script` once via `interpreter` (with `-u` for unbuffered stdout) and reads its stdout line by line. Each line is parsed as a `float` and applied as the new `target.demand`:

- unparsable lines are logged and skipped
- negative values are clamped to `0` (and logged)

The subprocess inherits cobald's `stderr`, so the script's tracebacks and stderr output show up directly in cobald's log.

When the streaming process crashes, it is made sure, that the subprocess is properly exited and terminated of not. When the script exits (crash or clean exit) it is restarted after `restart_delay` seconds (default `3`).

Example:

coabld_config.yml
```
 pipeline:
   - !StreamingController
     script: "/home/MyUser/run/streaming_script.py"                                                                                                                                                                   
     interpreter: "/home/MyUser/envs/snowflake311/bin/python"
```

streaming_script.py
```
 print("1")                 # demand 1
 print("-1")                # cobald.controller.streaming: 2026-07-31 13:00:48 received negative demand -1.0, clamping to 0
 print("aaa")             # cobald.controller.streaming: 2026-07-31 12:36:54 could not convert string to float: 'aaa'
 print(1)                    # demand 1
 print(None)             # cobald.controller.streaming: 2026-07-31 13:00:48 could not convert string to float: 'None'
  
 undefine_variable    # File "/home/MyUser/run/streaming_script.py", line 16, in <module>  \n undefine_variable   [file crashes -> restart]
```
This was also tested for a `shell` script.

Is mocking the subprocess via `FakeProcess` the right level here, or would you prefer a lighter/more detailed testing here?

### CODE REVIEW

File: cobald_tests/controller/test_streaming.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    return str(script)
+
+
+async def run_briefly(coro_factory, duration):
+    """Run a coroutine as a task, cancel it after ``duration``, and await it"""
+    task = asyncio.create_task(coro_factory())
+    await asyncio.sleep(duration)
+    task.cancel()
+    with contextlib.suppress(asyncio.CancelledError):
+        await task
```
[the remark this answers is not in the document]
Comment (lsowa) [reply cobald_tests/controller/test_streaming.py]: It does: the cancellation is only scheduled in the event loop. To execute it we have to hand over the control to the event loop again which we do by awaiting `task`. For testing we want this to happen here, within the `run_briefly` function.

[the remark this answers is not in the document]
Comment (maxfischer2781) [reply cobald_tests/controller/test_streaming.py]: This report is a false positive by CodeQL, see https://github.com/github/codeql/issues/11235. Please ignore it and anything like it.

File: cobald_tests/controller/test_streaming.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+from ..mock.pool import MockPool
+
+
+def write_script(tmp_path, body):
+    script = tmp_path / "script.py"
+    script.write_text(textwrap.dedent(body))
+    return str(script)
+
+
+async def run_briefly(coro_factory, duration):
```
Comment (maxfischer2781) [inline cobald_tests/controller/test_streaming.py]: `asyncio` already ships with tools for [timeouts](https://docs.python.org/3/library/asyncio-task.html#timeouts) you can probably directly use [`asyncio.wait_for`](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for) instead of this function.

File: src/cobald/controller/streaming.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+logger = logging.getLogger("cobald.controller.streaming")
+
+
+@service(flavour=asyncio)
+class StreamingController(Controller):
+    """
+    Controller that sets demand from a continuous stream of values emitted by an
+    external script
+
+    The ``script`` is repeatedly run as a subprocess via ``interpreter``.
```
Comment (maxfischer2781) [inline src/cobald/controller/streaming.py]: I highly recommend not to restrict this to an `interpreter` + `script` setup. Every script can already define its own interpreter and act as an executable. Simply take any executable, which covers both interpreter+script as well as compiled executables.
If you want to be fully general, take a `command: list[str]` similar to how `subprocess` itself does.

File: src/cobald/controller/streaming.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    Each line the subprocess writes to stdout is parsed as a ``float`` and
+    applied as the new demand of ``target``; lines that cannot be parsed are
+    logged and skipped. Negative values are clamped to ``0`` and logged as a
+    warning. If the subprocess exits or fails, it is restarted after waiting
+    ``restart_delay`` seconds.
```
Comment (maxfischer2781) [inline src/cobald/controller/streaming.py]: I recommend to make the error handling either always severe or at least configurable and default to severe. In most deployments there won't be anyone around to search through logs for warnings unless the entire service crashes.

File: src/cobald/controller/streaming.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    """
+
+    def __init__(
+        self, target: Pool, script: str, interpreter: str, restart_delay: float = 3
+    ):
+        super().__init__(target=target)
+
+        enforce(
+            restart_delay >= 0, InvariantError("restart_delay must not be negative")
+        )
```
Comment (maxfischer2781) [inline src/cobald/controller/streaming.py]: I would like to deprecate `enforce` with #170. Please use a regular `assert`.
```suggestion
        assert restart_delay >= 0, "restart_delay must not be negative"
```

File: cobald_tests/controller/test_streaming.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lsowa) [reply cobald_tests/controller/test_streaming.py]: Thank you for pointing this out!

File: src/cobald/controller/streaming.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            try:
+                await self._stream_demand()
+            except Exception:
+                logger.exception("streaming controller subprocess failed")
+            await asyncio.sleep(self.restart_delay)
+
+    async def _stream_demand(self):
+        proc = await asyncio.create_subprocess_exec(
+            self.interpreter,
+            "-u",
```
Comment (maxfischer2781) [inline src/cobald/controller/streaming.py]: Most interpreters or even executables won't understand [this flag](https://docs.python.org/3/using/cmdline.html#cmdoption-u). A safe support would be to set [the corresponding environment variable `PYTHONUNBUFFERED`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONUNBUFFERED), but this is only possible for Python (Ruby, JVM, bash, …, don't have equivalent env variable). You could also use [`unbuffer -p` or `stdbuf`](https://unix.stackexchange.com/questions/97294/unbuffer-or-stdbuf-for-removing-stdout-buffering) but that's not guaranteed to be available.

A simple way forward would be to document and expect scripts to be line- or unbuffered. People can write their program appropriately, or manually add `unbuffer`/`stdbuf` if available.

File: src/cobald/controller/streaming.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            # stream demand
+            async for line in proc.stdout:
+                try:
+                    demand = float(line.decode().strip())
+                except (ValueError, TypeError) as e:
+                    logger.warning(e)
+                    continue
+                if demand < 0:
+                    logger.warning("received negative demand %r, clamping to 0", demand)
+                    demand = 0
```
Comment (maxfischer2781) [inline src/cobald/controller/streaming.py]: See my leading comment on the class. I think you should just expect the scripts to work and not try to handle mistakes in a way that may be suboptimal.
```suggestion
                demand = float(line.decode().strip())
```

File: src/cobald/controller/streaming.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+from cobald.daemon import service
+from cobald.interfaces import Controller, Pool
+from cobald.utility import InvariantError, enforce
+
+logger = logging.getLogger("cobald.controller.streaming")
+
+
+@service(flavour=asyncio)
+class StreamingController(Controller):
```
Comment (maxfischer2781) [inline src/cobald/controller/streaming.py]: Please try to find a different name. Streaming is very general, the name should at least communicate that processes are involved. Something like `SubprocessController` could do.

**Review by maxfischer2781 — CHANGES_REQUESTED** [review]
Mostly comments on the implementation for now.

File: src/cobald/controller/streaming.py (hunk 8 above)
Comment (lsowa) [reply src/cobald/controller/streaming.py]: I renamed the controller to `SubprocessController`.

File: cobald_tests/controller/test_streaming.py (hunk 2 above)
Comment (lsowa) [reply cobald_tests/controller/test_streaming.py]: In fact, `async.wait_for` does exactly this... I dropped `run_briefly` and replaced it with `async.wait_for`.

File: src/cobald/controller/streaming.py (hunk 3 above)
Comment (lsowa) [reply src/cobald/controller/streaming.py]: You're right, the interpreter part is unnecessary overhead. I decided to go with you're second approach `command: list[str]` to be more general and allow for arguments.

File: src/cobald/controller/streaming.py (hunk 5 above)
Comment (lsowa) [reply src/cobald/controller/streaming.py]: used assert instead of enforce.

File: src/cobald/controller/streaming.py (hunk 7 above)
Comment (lsowa) [reply src/cobald/controller/streaming.py]: Implemented this part as proposed.

File: src/cobald/controller/streaming.py (hunk 6 above)
Comment (lsowa) [reply src/cobald/controller/streaming.py]: This is indeed a bug. I'll mention the buffering problem in the documentation. So the user can specify unbuffered `stdout` either in the scrip or in the way the script is called in the `cobald.yml` (e,g, `command: ["python", "-u", "script.py"]`).

File: src/cobald/controller/streaming.py (hunk 4 above)
Comment (lsowa) [reply src/cobald/controller/streaming.py]: I adapted the setup to call exceptions instead of warnings.

**Comment by lsowa** [comment]
Thanks for your helpful comments! I adapted my code accordingly and renamed the module to `SubprocessController`. If the code has green light i would adapt the tests.

File: src/cobald/controller/subprocess.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+    ``command`` must write unbuffered, or demand updates lag.
+
+    :param target: the pool to manage
+    :param command: command producing demand values on stdout, as a list of
+        the executable and its arguments, e.g. ``["python3", "-u", "script.py"]``
+    :param restart_delay: delay in seconds before restarting a stopped command
+    """
+
+    def __init__(self, target: Pool, command: List[str], restart_delay: float = 3):
```
Comment (maxfischer2781) [inline src/cobald/controller/subprocess.py]: Minor point (it's also fine not to cover in this PR) but we should also offer to *not* restart the process (and instead crash and burn everything).
```suggestion
    def __init__(self, target: Pool, command: List[str], restart_delay: "float | None" = 3):
```

File: src/cobald/controller/subprocess.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    async def _stream_demand(self):
+        proc = await asyncio.create_subprocess_exec(
+            *self.command,
+            stdout=asyncio.subprocess.PIPE,
+        )
+
+        try:
+            # stream demand
+            async for line in proc.stdout:
+                self.target.demand = float(line.decode().strip())
```
Comment (maxfischer2781) [inline src/cobald/controller/subprocess.py]: You don't need to decode or strip, `float` handles padded bytes just fine. Unicode specific numerals like `"²"` aren't converted by `float` either way.

```suggestion
                self.target.demand = float(line)
```

File: src/cobald/controller/subprocess.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+            # wait for process to finish
+            await proc.wait()
+        finally:
+            # in case of not finishing properly, ensure proper temrination
```
Comment (maxfischer2781) [inline src/cobald/controller/subprocess.py]: #NITPICK
```suggestion
            # in case of not finishing properly, ensure proper termination
```

**Review by maxfischer2781 — COMMENTED** [review]
Some minor comments, but please feel free to go ahead with tests.

**Comment by lsowa** [comment]
Okay, tests and pipelines are adapted and fixed.

File: cobald_tests/controller/test_subprocess.py
Comment (maxfischer2781) [inline cobald_tests/controller/test_subprocess.py:1]: Tests are looking fine to me.

Please also test the case that `restart_delay` is `None`.

File: src/cobald/controller/subprocess.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    :param target: the pool to manage
+    :param command: command producing demand values on stdout, as a list of
+        the executable and its arguments, e.g. ``["python3", "-u", "script.py"]``
+    :param restart_delay: delay in seconds before restarting a stopped command
+    """
+
+    def __init__(
+        self,
+        target: Pool,
+        command: List[str],
```
Comment (maxfischer2781) [inline src/cobald/controller/subprocess.py:34]: Since you are using quoted types anyways, please go for builtin types instead of `typing` wrappers wherever possible. You can drop the respective imports as well, then.

```suggestion
        command: "list[str]",
```

'zis iz ze futschah.

File: src/cobald/controller/subprocess.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+        restart_delay: "float | None" = 3,
+    ):
+        super().__init__(target=target)
+
+        assert restart_delay >= 0, "restart_delay must not be negative"
```
Comment (maxfischer2781) [inline src/cobald/controller/subprocess.py:39]: This would fail if `restart_delay` is `None` right now.
```suggestion
        assert restart_delay is None or restart_delay >= 0, "restart_delay must not be negative"
```

File: src/cobald/controller/subprocess.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+        self.command = command
+        self.restart_delay = restart_delay
+
+    async def run(self):
+        while True:
+            try:
+                await self._stream_demand()
+            except Exception:
+                logger.exception("subprocess controller subprocess failed")
```
Comment (maxfischer2781) [inline src/cobald/controller/subprocess.py:49]: At some point you need to bail out if no sleep is desired (`restart_delay = None`). I recommend to do it here, as then you can nicely propagate the error and don't have to special-case the `asyncio.sleep`.

```suggestion
                logger.exception("subprocess controller subprocess failed")
                if self.restart_delay is None:
                    raise
```

File: src/cobald/controller/subprocess.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+        proc = await asyncio.create_subprocess_exec(
+            *self.command,
+            stdout=asyncio.subprocess.PIPE,
+        )
+
+        try:
+            # stream demand
+            async for line in proc.stdout:
+                self.target.demand = float(line)
+            # wait for process to finish
```
Comment (maxfischer2781) [inline src/cobald/controller/subprocess.py:62]: Nit: Don't say what you do, say why you do it.

```suggestion
            # wait for process to clean up
```

File: src/cobald/controller/subprocess.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    Controller that sets demand from a continuous stream of values emitted by an
+    external command
+
+    The ``command`` is repeatedly run as a subprocess. Each line the subprocess
+    writes to stdout is parsed as a ``float`` and applied as the new demand of
+    ``target``. If the subprocess exits, fails to start, or writes a line that
+    cannot be parsed as a demand, the failure is logged and the command is
+    restarted after waiting ``restart_delay`` seconds.
+
+    ``command`` must write unbuffered, or demand updates lag.
```
Comment (maxfischer2781) [inline src/cobald/controller/subprocess.py:23]: Perhaps this is a bit easier to understand if one is not familiar with the topic.
```suggestion
    ``command`` must write to stdout unbuffered, otherwise demand updates may be delayed.
```

**Review by maxfischer2781 — CHANGES_REQUESTED** [review]
The `restart_delay = None` path needs some more handling, otherwise this is only nits for improving code a tiny bit while we are at it.

File: src/cobald/controller/subprocess.py (hunk 12 above)
Comment (lsowa) [reply src/cobald/controller/subprocess.py]: Oh this good :ok_hand:

File: cobald_tests/controller/test_subprocess.py
Comment (lsowa) [reply cobald_tests/controller/test_subprocess.py]: implemented testcase

File: src/cobald/controller/subprocess.py (hunk 13 above)
Comment (lsowa) [reply src/cobald/controller/subprocess.py]: added assert condition for `restart_delay=None`

File: src/cobald/controller/subprocess.py (hunk 14 above)
Comment (lsowa) [reply src/cobald/controller/subprocess.py]: added as proposed

File: src/cobald/controller/subprocess.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
+import asyncio
+import logging
+from typing import List
```
[the remark this answers is not in the document]
Comment (lsowa) [reply src/cobald/controller/subprocess.py]: fixed in [b00f58e](https://github.com/MatterMiners/cobald/commit/b00f58e0583a42c754e55792d15a7feccc20e6ec).

**Comment by lsowa** [comment]
Thank you for your comments! I implemented the handling of `restart_delay=None` and changes as requested.

**Review by maxfischer2781 — APPROVED** [review]
Looks good to me, ready for merging. 👍
